runtime/storage/cache/manager_internal.go

```diff
@@ -0,0 +1,431 @@
+package cache
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"math/rand"
+	"runtime"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/alicebob/miniredis/v2"
+	"github.com/go-redis/redis/v8"
+	jsoniter "github.com/json-iterator/go"
+
+	"encore.dev/appruntime/config"
+	"encore.dev/appruntime/reqtrack"
+	"encore.dev/appruntime/runtimeutil/syncutil"
+	"encore.dev/appruntime/testsupport"
+	"encore.dev/appruntime/trace"
+	"encore.dev/internal/stack"
+)
+
+// Manager manages cache clients.
+type Manager struct {
+	cfg  *config.Config
+	rt   *reqtrack.RequestTracker
+	ts   *testsupport.Manager
+	json jsoniter.API
+
+	initTestSrv syncutil.Once
+	testSrv     *miniredis.Miniredis
+
+	clientMu sync.RWMutex
+	clients  map[string]*redis.Client
+}
+
+func NewManager(cfg *config.Config, rt *reqtrack.RequestTracker, ts *testsupport.Manager, json jsoniter.API) *Manager {
+	return &Manager{
+		cfg:     cfg,
+		rt:      rt,
+		ts:      ts,
+		json:    json,
+		clients: make(map[string]*redis.Client),
+	}
+}
+
+func (mgr *Manager) getClient(clusterName string) *redis.Client {
+	mgr.clientMu.RLock()
+	cl := mgr.clients[clusterName]
+	mgr.clientMu.RUnlock()
+	if cl != nil {
+		return cl
+	}
+
+	// Client not found; acquire a write lock and set up the client
+	mgr.clientMu.Lock()
+	defer mgr.clientMu.Unlock()
+
+	// Did we race someone else and they set up the client first?
+	if cl := mgr.clients[clusterName]; cl != nil {
+		return cl
+	}
+
+	// Are we in a test or running in Encore Cloud? If so, use the redismock library.
+	if mgr.cfg.Static.Testing || mgr.runningInEncoreCloud() {
+		cl, err := mgr.newMiniredisClient()
+		if err != nil {
+			panic(fmt.Sprintf("cache: unable to start redis mock: %v", err))
+		}
+		mgr.clients[clusterName] = cl
+		return cl
+	}
+
+	for _, rdb := range mgr.cfg.Runtime.RedisDatabases {
+		if rdb.EncoreName == clusterName {
+			cl, err := mgr.newClient(rdb)
+			if err != nil {
+				panic(fmt.Sprintf("cache: unable to create redis client: %v", err))
+			}
+			mgr.clients[clusterName] = cl
+			return cl
+		}
+	}
+
+	panic(fmt.Sprintf("cache: unknown cluster %q", clusterName))
+}
+
+func (mgr *Manager) runningInEncoreCloud() bool {
```

Comment:
This feels like it should be controlled by config, rather than being hard coded like this.